File: ICDB/src/main/java/crypto/AlgorithmType.java
package crypto;

import crypto.signer.CmacSigner;
import crypto.signer.HmacSigner;
import crypto.signer.RsaSigner;
import crypto.signer.SipHashSigner;


/**
 * Created on 5/21/2016
 * @author Dan Kondratyuk
 */
public enum AlgorithmType {
    RSA {
        @Override
        public byte[] generateSignature(final byte[] data, final Key key) {
            return RsaSigner.generate(data, key.getPrivate());
        }

        @Override
        public boolean verify(final byte[] data, final Key key, final byte[] signature) {
            return RsaSigner.verify(data, key.getPublic(), signature);
        }
    },
    AES {
        @Override
        public byte[] generateSignature(final byte[] data, final Key key) {
            return CmacSigner.generate(data, key);
        }

        @Override
        public boolean verify(final byte[] data, final Key key, final byte[] signature) {
            return CmacSigner.verify(data, key, signature);
        }
    },
    SHA {
        @Override
        public byte[] generateSignature(final byte[] data, final Key key) {
            return HmacSigner.generate(data, key);
        }

        @Override
        public boolean verify(final byte[] data, final Key key, final byte[] signature) {
            return HmacSigner.verify(data, key, signature);
        }
    };

    public abstract byte[] generateSignature(final byte[] data, final Key key);
    public abstract boolean verify(final byte[] data, final Key key, final byte[] signature);
}

File: ICDB/src/main/java/crypto/AlgorithmType.kt
package crypto

import crypto.signer.MacSigner
import crypto.signer.RSASHA1Signer
import crypto.signer.RsaSigner
import main.args.config.UserConfig


/**
 * Enumerates all supported algorithm types, with extension methods to generate and verify signatures using its
 * corresponding algorithm implementation.
 *
 * Created on 5/21/2016
 * @author Dan Kondratyuk
 */
enum class AlgorithmType {
    RSA {
        override fun generateSignature(data: ByteArray, key: Key) =
         RsaSigner.generate(data, key.publicRsaKey)

        override fun verify(data: ByteArray, key: Key, signature: ByteArray) =
            RsaSigner.verify(data, key.privateRsaKey, signature)
    },
    RSA_AGGREGATE {
        override fun generateSignature(data: ByteArray, key: Key) =
                RSASHA1Signer(key.modulus,key.exponent).computeSHA1RSA(data)


        override fun verify(data: ByteArray, key: Key, signature: ByteArray) =
                RsaSigner.verify(data, key.privateRsaKey, signature)
    },
    AES {
        override fun generateSignature(data: ByteArray, key: Key) =
            MacSigner.generate(data, key, MacSigner.cmacAes)

        override fun verify(data: ByteArray, key: Key, signature: ByteArray) =
            MacSigner.verify(data, key, signature, MacSigner.cmacAes)
    },
    SHA {
        override fun generateSignature(data: ByteArray, key: Key) =
            MacSigner.generate(data, key, MacSigner.hmacSha)

        override fun verify(data: ByteArray, key: Key, signature: ByteArray) =
            MacSigner.verify(data, key, signature, MacSigner.hmacSha)
    };

    abstract fun generateSignature(data: ByteArray, key: Key): ByteArray
    abstract fun verify(data: ByteArray, key: Key, signature: ByteArray): Boolean
}
